src/api/uploads/MultiputUpload.js

```diff
@@ -0,0 +1,358 @@
+        if (this.isDestroyed()) {
+            return;
+        }
+
+        const { id, part_size, session_endpoints } = data;
+
+        this.sessionId = id;
+        this.partSize = part_size;
+        this.sessionEndpoints = {
+            ...this.sessionEndpoints,
+            uploadPart: session_endpoints.upload_part,
+            listParts: session_endpoints.list_parts,
+            commit: session_endpoints.commit,
+            abort: session_endpoints.abort,
+            logEvent: session_endpoints.log_event
+        };
+
+        this.populateParts();
+        this.processNextParts();
+    };
+
+    /**
+     * Attempts to process more parts, except in the case where everything is done or we detect
+     * a file change (in which case we want to abort and not process more parts).
+     * 
+     * @private
+     * @return {void}
+     */
+    processNextParts = (): void => {
+        if (this.failSessionIfFileChangeDetected()) {
+            return;
+        }
+
+        if (this.partsUploaded === this.parts.length && this.fileSha1) {
+            this.commitSession();
+            return;
+        }
+
+        this.updateFirstUnuploadedPartIndex();
+
+        while (this.canStartMorePartUploads()) {
+            this.uploadNextPart();
+        }
+
+        // TODO: compute digests for parts
+    };
+
+    /**
+     * Send a request to commit the upload.
+     * TODO: implement this
+     * 
+     * @private
+     * @return {void}
+     */
+    commitSession = (): void => {};
+
+    /**
+     * Commit response handler.  Succeeds the upload, retries the commit on 202/429/failure, or fails
+     * the upload.
+     * TODO: implement this
+     * 
+     * @private
+     * @return {void}
+     */
+    commitSessionSuccessHandler = (): void => {};
+
+    /**
+     * Commit error handler.
+     * Retries the commit or fails the multiput session.
+     * TODO: implement this
+     * 
+     * @private
+     * @return {void}
+     */
+    commitSessionErrorHandler = (): void => {};
+
+    /**
+     * Find first part in parts array that we can upload, and upload it.
+     * 
+     * @private
+     * @return {void}
+     */
+    uploadNextPart = (): void => {
+        for (let i = this.firstUnuploadedPartIndex; i < this.parts.length; i += 1) {
+            const part = this.parts[i];
+
+            if (part.state === PART_STATE_DIGEST_READY) {
+                // Update the counters here instead of uploadPart because uploadPart
+                // can get called on retries
+                this.partsDigestReady -= 1;
+                this.partsUploading += 1;
+                part.upload();
+                break;
+            }
+        }
+    };
+
+    /**
+     * Checks if upload pipeline is full
+     * TODO: support progressive hashing
+     * 
+     * @private
+     * @return {boolean}
+     */
+    canStartMorePartUploads = (): boolean => !this.isDestroyed() && this.partsUploading < this.config.parallelism;
```

Comment:
We should probably play with heavily boosting the parallelism @klk-. At least Chrome supports a pretty high number.